mdp/linear_flows_online.py

```diff
@@ -0,0 +1,607 @@
+        the last node has no multiple training phases) that must return data
+        arrays to train nodes (so the data arrays are the 'x' for the nodes).
+        Note that the data arrays are processed by the nodes
+        which are in front of the node that gets trained,
+        so the data dimension must match the input dimension of the first node.
+
+        'data_iterables' can also be a 2D or a 3D numpy array. A 2D array trains
+        all the nodes incrementally, while a 3D array supports online training
+        in batches (=shape[1]).
+
+        'data_iterables' can also return a list or a tuple, where the first entry is
+        'x' and the following are the required args for training all the nodes in
+        the flow (e.g. for supervised training).
+
+        (x, (node-0 args), (node-1 args), ..., (node-n args)) - args for n nodes
+
+        if say node-i does not require any args, the provided (node-i args) are ignored.
+        So, one can simply use None for the nodes that do not require args.
+
+        (x, (node-0 args), ..., None, ..., (node-n args)) - No args for the ith node.
+
+        """
+
+        data_iterables = self._train_check_iterables(data_iterables)
+
+        if self.verbose:
+            strn = [str(self.flow[i]) for i in xrange(len(self.flow))]
+            print("Training nodes %s simultaneously" % strn)
+        self._train_nodes(data_iterables)
+
+        # close training the terminal node only if it is a trainable Node
+        if not isinstance(self.flow[-1], mdp.OnlineNode):
+            self._close_last_node()
+
+    # private container methods
+
+    def _check_value_type_is_compatible(self, value):
+        # onlinenodes, trained and non-trainable nodes are compatible
+        if not isinstance(value, mdp.Node):
+            raise TypeError("flow item must be a Node instance and not %s" % type(value))
+        elif isinstance(value, mdp.OnlineNode):
+            pass
+        else:
+            # classic mdp Node
+            if value.is_training():
+                raise TypeError("flow item must either be an OnlineNode instance, a trained or a non-trainable Node.")
+
+    def _check_compatibilitiy(self, flow):
+        [self._check_value_type_is_compatible(item) for item in flow[:-1]]
+        # terminal node can be a trainable Node
+        self._check_value_type_isnode(flow[-1])
+
+    def _get_cache_from_flow(self, flow):
+        _cache = {}
+        for i, node in enumerate(flow):
+            if not hasattr(node, 'cache'):
+                _cache['node#%d' % i] = {}
+            else:
+                _cache['node#%d' % i] = node.cache
+        return _cache
+
+    def __setitem__(self, key, value):
+        if isinstance(key, slice):
+            [self._check_value_type_is_compatible(item) for item in value]
+        else:
+            self._check_value_type_is_compatible(value)
+
+        # make a copy of list
```

Comment:
I'd leave out comments like these that just describe what the code is doing (without providing real value).